File: packages/setpieces-r/setpieces/R/teams.R
# Convenience helper to map team names to Opta contestantIds from this
# repo's match-file naming convention: "YYYY-MM-DD_Home Team - Away Team.json".
#
# This is optional -- every other function in the package works
# directly off contestantId strings, so use this only if your file
# layout follows that convention.

FILENAME_RE <- "^\\d{4}-\\d{2}-\\d{2}_(.+) - (.+)\\.json$"

#' Infer team name -> contestantId from match filenames
#'
#' A team name resolves only if the intersection of contestantIds seen
#' across every match file where that team appears (as home or away)
#' narrows to exactly one ID.
#'
#' @param paths Character vector of match JSON file paths, named
#'   `YYYY-MM-DD_Home Team - Away Team.json`.
#' @return A named list, team display name -> contestantId.
#' @export
team_ids_from_filenames <- function(paths) {
  team_cid_sets <- list()

  for (path in paths) {
    base <- basename(path)
    m <- regmatches(base, regexec(FILENAME_RE, base))[[1]]
    if (length(m) < 3) next
    home <- m[2]
    away <- m[3]

    match <- load_match(path)
    cids <- unique(vapply(match$event, function(e) {
      if (is.null(e$contestantId)) NA_character_ else e$contestantId
    }, character(1)))
    cids <- cids[!is.na(cids)]

    for (team in c(home, away)) {
      if (is.null(team_cid_sets[[team]])) team_cid_sets[[team]] <- list()
      team_cid_sets[[team]][[length(team_cid_sets[[team]]) + 1]] <- cids
    }
  }

  result <- list()
  for (team in names(team_cid_sets)) {
    sets <- team_cid_sets[[team]]
    inter <- Reduce(intersect, sets)
    if (length(inter) == 1) result[[team]] <- inter[1]
  }
  result
}

File: packages/setpieces-py/src/setpieces/teams.py
"""Convenience helper to map team names to Opta contestantIds from this
repo's match-file naming convention: ``YYYY-MM-DD_Home Team - Away Team.json``.

This is optional -- every other function in the package works directly
off ``contestantId`` strings, so use this only if your file layout
follows that convention.
"""

import collections
import json
import re
from typing import Dict, List

FILENAME_RE = re.compile(r"\d{4}-\d{2}-\d{2}_(.+) - (.+)\.json$")


def team_ids_from_filenames(paths: List[str]) -> Dict[str, str]:
    """team display name -> contestantId, inferred by intersecting the
    contestantIds seen in every match file where that team's name
    appears as the home or away side."""
    team_cid_sets = collections.defaultdict(list)
    for path in paths:
        m = FILENAME_RE.search(path.split("/")[-1])
        if not m:
            continue
        home, away = m.group(1), m.group(2)
        with open(path, encoding="utf-8") as f:
            data = json.load(f)
        cids = {e["contestantId"] for e in data.get("event", []) if "contestantId" in e}
        team_cid_sets[home].append(cids)
        team_cid_sets[away].append(cids)

    team_to_cid = {}
    for team, sets in team_cid_sets.items():
        inter = set.intersection(*sets)
        if len(inter) == 1:
            team_to_cid[team] = next(iter(inter))
    return team_to_cid
